File: Protocol/Responses/GetContactsResponse.cpp
#include "GetContactsResponse.h"

GetContactsResponse::GetContactsResponse(vector<ContactViewModel> contacts)
{
    this->contacts = contacts;
}

vector<ContactViewModel> GetContactsResponse::GetContacts() const
{
    return this->contacts;
}

ResponseType GetContactsResponse::GetResponseType() const
{
    return ResponseType::GetContacts;
}

string GetContactsResponse::ToString() const
{
    ostringstream stream;

    int responseType = (int)GetResponseType();
    stream.write((char *)&responseType, sizeof(responseType));

    int count = contacts.size();
    stream.write((char *)&count, sizeof(count));

    for (int i = 0; i < count; i++)
    {
        int length = contacts[i].ContactName.size();
        stream.write((char *)&length, sizeof(length));
        stream.write(&contacts[i].ContactName[0], length);
        stream.write((char *)&contacts[i].IsOnline, sizeof(contacts[i].IsOnline));
    }

    return stream.str();
}

File: Client/Protocol/Responses/GetContactsResponse.cs
using NetTalk.Model;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace NetTalk.Protocol.Responses
{
    public class GetContactsResponse : ResponseBase
    {
        public GetContactsResponse(List<Contact> contacts)
        {
            Contacts = contacts;
        }

        public List<Contact> Contacts { get; }
        public override ResponseType ResponseType { get; } = ResponseType.AddContact;

        public static GetContactsResponse Parse(byte[] bytes)
        {
            using (MemoryStream stream = new MemoryStream(bytes))
            using (BinaryReader reader = new BinaryReader(stream))
            {
                ResponseType responseType = (ResponseType)reader.ReadInt32();
                if (responseType != ResponseType.GetContacts)
                    throw new InvalidOperationException("Invalid type of response");

                int contactCount = reader.ReadInt32();
                List<Contact> contacts = new List<Contact>();
                for (int i = 0; i < contactCount; i++)
                {
                    Contact contact = new Contact();

                    contact.Name = Encoding.ASCII.GetString(reader.ReadBytes(reader.ReadInt32()));
                    contact.IsOnline = reader.ReadBoolean();

                    contacts.Add(contact);
                }

                return new GetContactsResponse(contacts);
            }
        }
    }
}